pkg/gardenlet/controller/managedseed/managedseed_reconciler.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	if managedSeed.DeletionTimestamp != nil {
+		return r.delete(gardenClient, managedSeed)
+	}
+	return r.createOrUpdate(gardenClient, managedSeed)
+}
+
+func (r *reconciler) createOrUpdate(gardenClient kubernetes.Interface, managedSeed *seedmanagementv1alpha1.ManagedSeed) (reconcile.Result, error) {
+	managedSeedLogger := logger.NewFieldLogger(r.logger, "managedSeed", kutil.ObjectName(managedSeed))
+
+	// Ensure gardener finalizer
+	if err := controllerutils.EnsureFinalizer(r.ctx, gardenClient.DirectClient(), managedSeed, gardencorev1beta1.GardenerName); err != nil {
+		managedSeedLogger.Errorf("Could not ensure gardener finalizer on ManagedSeed: %+v", err)
+		return reconcile.Result{}, err
+	}
+
+	// Get shoot
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := gardenClient.Client().Get(r.ctx, kutil.Key(managedSeed.Namespace, managedSeed.Spec.Shoot.Name), shoot); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if the shoot is reconciled
+	if shoot.Generation != shoot.Status.ObservedGeneration || shoot.Status.LastOperation == nil || shoot.Status.LastOperation.State != gardencorev1beta1.LastOperationStateSucceeded {
+		managedSeedLogger.Infof("Waiting for shoot %s to be reconciled before registering it as seed", kutil.ObjectName(shoot))
+
+		// Update status to Pending
+		if updateErr := updateStatusPending(r.ctx, gardenClient.DirectClient(), managedSeed, "Reconciliation pending"); updateErr != nil {
+			managedSeedLogger.Errorf("Could not update ManagedSeed status for pending creation or update: %+v", updateErr)
+			return reconcile.Result{}, updateErr
+		}
+
+		// Return success result with requeue after 10 seconds
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	// Get seed client
+	seedClient, err := r.clientMap.GetClient(r.ctx, keys.ForSeedWithName(*shoot.Spec.SeedName))
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get seed client: %w", err)
+	}
+
+	// Get shoot client
+	shootClient, err := r.clientMap.GetClient(r.ctx, keys.ForShoot(shoot))
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get shoot client: %w", err)
+	}
+
+	// Create actuator
+	a := newActuator(gardenClient, seedClient, shootClient, r.config, r.imageVector, r.logger)
+
+	// Update status to Processing
+	if updateErr := updateStatusProcessing(r.ctx, gardenClient.DirectClient(), managedSeed, "Reconciling"); updateErr != nil {
+		managedSeedLogger.Errorf("Could not update ManagedSeed status before creation or update: %+v", updateErr)
+		return reconcile.Result{}, updateErr
+	}
+
+	// Reconcile creation or update
+	managedSeedLogger.Debug("Reconciling ManagedSeed creation or update")
+	if err := a.CreateOrUpdate(r.ctx, managedSeed, shoot); err != nil {
+		managedSeedLogger.Errorf("Could not reconcile ManagedSeed creation or update: %+v", err)
+
+		// Record an event
+		lastError := &gardencorev1beta1.LastError{
+			Codes:       gardencorev1beta1helper.ExtractErrorCodes(err),
```

Comment:
I don't see any usage of error codes in the actuator, which could be extracted here.

Reply:
There is none. Isn't `LastError` supposed to provide info to the human user as well?